integration/src/builtin_functions.rs

```diff
@@ -0,0 +1,139 @@
+    params: &[ValType::I32, ValType::I32, ValType::I32],
+    results: &[],
+};
+
+const MEMORY_FILL_WASM: BuiltinDefinition = BuiltinDefinition {
+    wasm_bytes: include_bytes!(env!("MEMORY_FILL_WASM_PATH")),
+    params: &[ValType::I32, ValType::I32, ValType::I32],
+    results: &[],
+};
+
+struct BuiltinFunction<F: PrimeField32> {
+    function_index: u32,
+    function_definition: WriteOnceAsm<Directive<F>>,
+}
+
+/// Tracks which built-in functions are used and provides functions to replace unsupported
+/// instructions with calls to these built-in functions.
+pub struct Tracker<F: PrimeField32> {
+    used_builtins: HashMap<*const u8, BuiltinFunction<F>>,
+}
+
+impl<F: PrimeField32> Tracker<F> {
+    pub fn new() -> Self {
+        Self {
+            used_builtins: HashMap::new(),
+        }
+    }
+
+    /// If the given instruction is not supported natively by the VM, replace it with a call to
+    /// a built-in function.
+    pub fn replace_with_builtins(
+        &mut self,
+        module: &mut Module,
+        label_gen: &mut RangeFrom<u32>,
+        dag: &mut BlocklessDag,
+    ) {
+        for node in &mut dag.nodes {
+            match &mut node.operation {
+                Operation::WASMOp(op) => {
+                    let builtin = match op {
+                        Operator::MemoryCopy { dst_mem, src_mem } => {
+                            // We don't support multi-memory modules.
+                            assert_eq!(*dst_mem, 0);
+                            assert_eq!(*src_mem, 0);
+                            &MEMORY_COPY_WASM
+                        }
+                        Operator::MemoryFill { mem } => {
+                            // We don't support multi-memory modules.
+                            assert_eq!(*mem, 0);
+                            &MEMORY_FILL_WASM
+                        }
+                        // TODO: Add more built-in functions here as needed.
+                        _ => continue,
+                    };
+
+                    // Get the index of the built-in function, adding it to the module if needed.
+                    let function_index = self
+                        .get_builtin_function(module, label_gen, builtin)
+                        .function_index;
+                    // Replace the instruction with a call to the built-in function.
+                    *op = Operator::Call { function_index };
+                }
+                Operation::Loop { sub_dag, .. } => {
+                    self.replace_with_builtins(module, label_gen, sub_dag);
+                }
+                _ => (),
+            }
+        }
+    }
+
+    /// Returns an iterator over the used built-in functions to be added to the final program.
+    /// Sorted by function index.
+    pub fn into_used_builtins(self) -> impl Iterator<Item = WriteOnceAsm<Directive<F>>> {
+        // We only support one built-in function for now.
+        assert!(self.used_builtins.len() <= 2);
```

Comment:
i don't get the comment and assert

Reply:
I was afraid each call site would append a copy of the function.